File: benchmarks/objectteams/app/src/BankBenchmark.java
import bank.Account;
import bank.Bank;
import bank.Person;
import bank.CallinTransaction;
import bank.CalloutTransaction;

public class BankBenchmark extends Benchmark {

	private Bank bank;
	private String optionalArgument;

	@Override
	public boolean innerBenchmarkLoop(final int innerIterations) {
		bank.activate();
		for (Account from : bank.getCheckingAccounts()) {
			float amount = from.getBalance() / (float) innerIterations;
			for (Account to : bank.getSavingAccounts()) {
				if(optionalArgument.equals("CALLIN")) {
					CallinTransaction transaction = new CallinTransaction();
					transaction.activate();
					try {
						transaction.execute(from, to, amount);
					} catch (RuntimeException e) {
						// do nothing
					} finally {
						transaction.deactivate();
					}
				} else {
					CalloutTransaction transaction = new CalloutTransaction();
					transaction.activate();
					try {
						transaction.execute(from, to, amount);
					} catch (RuntimeException e) {
						// do nothing
					} finally {
						transaction.deactivate();
					}
				}
			}
		}
		bank.deactivate();
		return true;
	}

	public boolean setUp(final int innerIterations) {
		bank = new Bank();
		bank.activate();

		for (int i = 0; i < innerIterations; ++i) {
			Person p = new Person();
			bank.addCustomer(p);

			Account sa = new Account(1000.0f);
			Account ca = new Account(1000.0f);
			bank.addSavingsAccount(p, sa);
			bank.addCheckingsAccount(p, ca);
		}
		return true;
	}

	public boolean setUpOptional(String optionalArgument) {
		this.optionalArgument = optionalArgument;
		return true;
	}

	@Override
	public Object benchmark() {
		throw new RuntimeException("Should never be reached");
	}

	@Override
	public boolean verifyResult(Object result) {
		throw new RuntimeException("Should never be reached");
	}
}

File: benchmarks/scroll/app/src/main/scala/BankBenchmark.scala
import bank._

import scroll.internal.support.DispatchQuery
import DispatchQuery._

class BankBenchmark extends Benchmark {

	var bank_var = new Bank

	override def innerBenchmarkLoop(innerIterations : Int) : Boolean = {
		val bank = bank_var
		
		for (from <- bank.getCheckingAccounts) {
			var amount : Float = from.getBalance / innerIterations;
			for (to <- bank.getSavingAccounts) {
				val transaction = new Transaction
				transaction.addPlaysRelation(from, new transaction.Source)
				transaction.addPlaysRelation(to, new transaction.Target)

				transaction partOf bank

				val transactionRole = new bank.TransactionRole
				bank.addPlaysRelation(transaction, transactionRole)
				transactionRole.execute(amount)
			}
		}
		return true;
	}

	def setUp(innerIterations : Int) : Boolean = {
			bank_var = new Bank
			val bank = bank_var

			for (i <- 1 to innerIterations) {
				var p = new Person
				bank.addPlaysRelation(p, new bank.Customer)

			 	var sa = new Account(1000.0f)
				var ca = new Account(1000.0f)

				bank.addPlaysRelation(sa, new bank.SavingsAccount)
				bank.addPlaysRelation(ca, new bank.CheckingsAccount)

				bank.addSavingsAccount(sa)
				bank.addCheckingsAccount(ca)
			}
		return true;
	}

	override def benchmark() : Object = {
		throw new RuntimeException("Should never be reached");
	}

	override def verifyResult(result : Object) : Boolean = {
		throw new RuntimeException("Should never be reached");
	}
}
